### PR TITLE
Orders

### PR DESCRIPTION
**Pull request by wezymord**

### CODE REVIEW

File: online_shop/admin.py [hunk 1]
```diff
@@ -24,11 +24,27 @@ class PhotoAdmin(admin.ModelAdmin):
...
     list_display = ['first_name', 'last_name', 'phone_number', 'e_mail']
-    list_filter = ['first_name', 'last_name', 'phone_number', 'e_mail']
 admin.site.register(User, UserAdmin)
 
 
+def shipping_methods(obj):
+    shipping_options = []
+    for option in obj.shipping_options.all():
+        shipping_options.append(option.shipping_method)
+    return ' - '.join(shipping_options)
```
Comment (pawel-krysiak) [inline online_shop/admin.py]:
```
shipping_options = [option.shipping_method for option in obj.shipping_options.all()]
' - '.join(shipping_options)
```
 is more elegant way. Read about list comprehensions in python :) 👍

File: online_shop/admin.py [hunk 2]
```diff
@@ -24,11 +24,27 @@ class PhotoAdmin(admin.ModelAdmin):
...
+
+def product_name(obj):
+    products = []
+    for product in obj.products.all():
+        products.append(product.name)
+    return ' - '.join(products)
```
Comment (pawel-krysiak) [inline online_shop/admin.py]: Same here. :) list comprehensions :)

File: online_shop/migrations/0001_initial.py [hunk 3]
```diff
@@ -13,29 +14,42 @@ class Migration(migrations.Migration):
 
     operations = [
         migrations.CreateModel(
-            name='Orders',
```
Comment (pawel-krysiak) [inline online_shop/migrations/0001_initial.py]: 👍

File: online_shop/templates/basket.html [hunk 4]
```diff
@@ -57,7 +57,7 @@ <h4 class="product-title"><a href="/product_details/{{ product.id }}">{{ product
 
                        {% for num in product.stock|get_stock_product_to_basket %}
 
-                           {% if num == products_amount|get_price_amount_product:product.id %}
+                           {% if num == products_amount|get_product_amount:product.id %}
```
Comment (pawel-krysiak) [inline online_shop/templates/basket.html]: You mean the quantity of the product? If yes maybe `product_quantity` would be better name?

File: online_shop/views.py [hunk 5]
```diff
@@ -120,5 +120,46 @@ def get(self, request):
         }
         return render(request, 'shop-grid-ns.html', ctx)
 
+class CheckoutAddress(View):
+    def get(self, request):
+        pass
```
Comment (pawel-krysiak) [inline online_shop/views.py]: Why `pass` not just render?

File: online_shop/views.py [hunk 6]
```diff
@@ -120,5 +120,46 @@ def get(self, request):
...
+        }
 
+        return render(request, 'checkout-shipping.html', ctx)
 
+
+class CheckoutReview(View):
+    def get(self, request):
+        products = []
+        if 'basket' in request.session.keys():
+            products_id = request.session['basket']
```
Comment (pawel-krysiak) [inline online_shop/views.py]: you mean `product_ids` ? :)

File: online_shop/views.py [hunk 7]
```diff
@@ -120,5 +120,46 @@ def get(self, request):
...
+                'products': list(set(products)),
+                'products_amount': request.session['basket']
+            }
+
+            return render(request, 'checkout-review.html', ctx)
+        else:
+            return render(request, 'checkout-review.html')
+
+    def post(self, request):
+        pass
```
Comment (pawel-krysiak) [inline online_shop/views.py]: the same question here and below, why `pass` ? not just render?

File: online_shop/views.py [hunk 8]
```diff
@@ -120,5 +120,46 @@ def get(self, request):
...
+class CheckoutReview(View):
+    def get(self, request):
+        products = []
+        if 'basket' in request.session.keys():
+            products_id = request.session['basket']
+            for id in products_id:
+                products.append(Product.objects.get(pk=id))
+
+            ctx = {
+                'products': list(set(products)),
```
Comment (pawel-krysiak) [inline online_shop/views.py]: You make set here. Does it mean that I can add to the session 1000 the same ids if I click very fast?

File: online_shop/views.py [hunk 9]
```diff
@@ -40,7 +40,7 @@ def get(self, request):
 class Basket(View):
     def get(self, request):
         products = []
-        if 'basket' in request.session.keys():
+        if 'basket' in request.session.keys():                       # zapisać products_amount w context processor (cena się wyświetla w buttonie koszyka)
```
Comment (pawel-krysiak) [inline online_shop/views.py]: @wezymord What is this comment about? is it working?

File: online_shop/views.py [hunk 10]
```diff
@@ -136,13 +140,19 @@ def get(self, request):
...
 
 class CheckoutReview(View):
     def get(self, request):
         products = []
         if 'basket' in request.session.keys():
-            products_id = request.session['basket']
-            for id in products_id:
+            product_ids = request.session['basket']
+            for id in product_ids:
                 products.append(Product.objects.get(pk=id))
```
Comment (pawel-krysiak) [inline online_shop/views.py]: You can do better :)
```
for product_id in request.session['basket']:
    products.append(Product.objects.get(pk=product_id))
```

**Comment by pawel-krysiak** [comment]
@wezymord Where are forms with the post method we talked about? ( the more simple way ) :)

File: online_shop/views.py (hunk 9 above)
Comment (wezymord) [reply online_shop/views.py]: Works, but I plan to improve it. I want return the total price in context processor for each page (base.html).

**Comment by wezymord** [comment]
The forms are in html, I return them in basket.js :) You didnt mean it?

**Comment by pawel-krysiak** [comment]
@wezymord nope. We don't submit forms via js in classic Web applications. Form url and button type submit should do the job :)

File: online_shop/views.py (hunk 9 above)
Comment (pawel-krysiak) [reply online_shop/views.py]: If you think that this is a better solution, and makes the solution more simple.. than sure.

File: online_shop/views.py [hunk 11]
```diff
@@ -120,5 +119,81 @@ def get(self, request):
         }
         return render(request, 'shop-grid-ns.html', ctx)
 
+class CheckoutAddress(View):
+    def get(self, request):
+        return render(request, 'checkout-address.html')
+
+    def post(self, request):
+        if request.method == 'POST':
```
Comment (pawel-krysiak) [inline online_shop/views.py]: @wezymord This does not automatically recognize the method? You need to make this check? I thought that this method automatically are assigned to specific request type Get/Post/Delete etc.

File: online_shop/views.py [hunk 12]
```diff
@@ -120,5 +119,81 @@ def get(self, request):
...
+        shipping_options = ShippingOption.objects.all()
+
+        ctx = {
+            'shipping_options': shipping_options
+        }
+
+        return render(request, 'checkout-shipping.html', ctx)
+
+    def post(self, request):
+        if request.method == 'POST':
```
Comment (pawel-krysiak) [inline online_shop/views.py]: @wezymord Same here.

File: static/basket_detail/basket.js [hunk 13]
```diff
@@ -107,4 +110,15 @@ $(document).ready(function() {
             }
         });
     });
+
+    $(".shipping-method").focus(function() {
+
+        $.ajax({
+            type: "POST",
+            url: "/checkout_shipping/",
```
Comment (pawel-krysiak) [inline static/basket_detail/basket.js]: @wezymord This dude does what? ;)

File: online_shop/views.py (hunk 11 above)
Comment (wezymord) [reply online_shop/views.py]: Yep, you're right

File: static/basket_detail/basket.js (hunk 13 above)
Comment (wezymord) [reply static/basket_detail/basket.js]: request.POST with shipping_method_id. You mean it should only be on back-end? ;)

**Pull request merged**